### PR TITLE
Add removeFiles and sortFiles actions

### PR DESCRIPTION
**Pull request by scott1702**
The changes made in this PR are to fix pagination, sorting and folder navigation which were previously broken.

### CODE REVIEW

**Comment by chillu** [comment]
Can you please rebase?

File: javascript/src/state/gallery/reducer.js [hunk 1]
```diff
@@ -30,11 +30,67 @@ export default function galleryReducer(state = initialState, action) {
 
     switch (action.type) {
         case GALLERY.ADD_FILE:
+            let nextFilesState = []; // Clone the state.files array
+
+            if (Object.prototype.toString.call(action.payload.file) === '[object Array]') {
```
Comment (chillu) [inline javascript/src/state/gallery/reducer.js]: Just a general note, we should start using lodash for this stuff

File: javascript/src/state/gallery/reducer.js [hunk 2]
```diff
@@ -30,11 +30,67 @@ export default function galleryReducer(state = initialState, action) {
 
     switch (action.type) {
         case GALLERY.ADD_FILE:
+            let nextFilesState = []; // Clone the state.files array
```
Comment (chillu) [inline javascript/src/state/gallery/reducer.js]: Why do we need both object and array variations in here? Don't we control the whole API usage and can simplify this?

File: javascript/src/state/gallery/reducer.js [hunk 3]
```diff
@@ -30,11 +30,67 @@ export default function galleryReducer(state = initialState, action) {
...
-                files: state.files.concat(action.payload.file)
+                count: typeof action.payload.count !== 'undefined' ? action.payload.count : state.count,
+                files: state.files.concat(nextFilesState)
             }));
 
+        case GALLERY.REMOVE_FILE:
+            if (typeof action.payload.id === 'undefined') {
+                // No param was passed, remove everything.
+                nextState = deepFreeze(Object.assign({}, state, { count: 0, files: [] }));
+            } else if (typeof action.payload.id === 'number') {
```
Comment (chillu) [inline javascript/src/state/gallery/reducer.js]: Same here, can't we just standardise on `ids` always being an array? This seems like a lot of duplication on the long term - Redux already creates a *lot* of boilerplate code with this :)

File: javascript/src/state/gallery/reducer.js (hunk 1 above)
Comment (flashbackzoo) [inline javascript/src/state/gallery/reducer.js]: Array.isArray() is legit got IE9 and above. And proper browsers.

File: javascript/src/state/gallery/actions.js [hunk 4]
```diff
@@ -1,16 +1,30 @@
...
  * @param number [count] - The number of files in the current view.
  */
-export function addFile(file, count) {
+export function addFiles(files, count) {
     return (dispatch, getState) => {
         return dispatch ({
-            type: GALLERY.ADD_FILE,
-            payload: { file, count }
+            type: GALLERY.ADD_FILES,
+            payload: { files, count }
```
Comment (chillu) [inline javascript/src/state/gallery/actions.js]: Hah, didn't know you can do that (object literal property value shorthand)

**Comment by chillu** [comment]
Great job, it's a lot cleaner now. Thanks for writing tests as well :)

File: javascript/src/state/gallery/actions.js (hunk 4 above)
Comment (scott1702) [inline javascript/src/state/gallery/actions.js]: ES6 :+1:

**Pull request merged**